include/arch/arm/aarch32/cortex_m/nvic.h

```diff
@@ -7,6 +7,8 @@
 #ifndef ZEPHYR_INCLUDE_ARCH_ARM_AARCH32_CORTEX_M_NVIC_H_
 #define ZEPHYR_INCLUDE_ARCH_ARM_AARCH32_CORTEX_M_NVIC_H_
 
+#include <devicetree.h>
```

Comment:
Why do we need to include devicetree.h here? Where is it explicitly required?

Reply:
If you take a look a few lines below:
```
#define NVIC_NODEID DT_INST(0, arm_v8m_nvic)
#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
#define NVIC_NODEID DT_INST(0, arm_v7m_nvic)
#elif defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
#define NVIC_NODEID DT_INST(0, arm_v6m_nvic)
#endif
#define NUM_IRQ_PRIO_BITS DT_PROP(NVIC_NODEID, arm_num_irq_priority_bits)
```

Reply:
I will need to understand why this inclusion is required now, and not when this header was introduced.

Reply:
Many SoCs tend to include `devicetree.h` somewhere along their headers, but that is apparently not the case for SAM E5x.

Either way, as long as the header makes use of device tree macros which are defined by `devicetree.h`, it should be including the `devicetree.h`.

Reply:
> Many SoCs tend to include devicetree.h somewhere along their headers, but that is apparently not the case for SAM E5x.

Then I would insist on having SAM E5x SoCs include devicetree.h in soc.h. We started introducing this convention a while back; most of ARM SoCs already have transitioned to include devicetree.h in their soc.h. We should do it for the few ARM SoCs that are remaining.

(Having this kind of consistent behavior across ARM SoCs helps very much in identifying inclusion cycles - something we have troubled a lot with - in ARM builds and makes it easier to do reviews of SoC definitions across different ARM vendors)

Regarding nvic.h: this header was introduced very recently, and is included in just two places, cmsis.h and exc.h and _both_ should be including devicetree.h before including nvic.h, so you don't need to include devicetree.h here (although not an error or a bad practice, this would help finding inconsistencies in the code, such as the mising devicetree inclusion in E5x socs.).

Reply:
> Regarding nvic.h: this header was introduced very recently, and is included in just two places, cmsis.h and exc.h and _both_ should be including devicetree.h before including nvic.h, so you don't need to include devicetree.h here

For `exc.h`, it includes `devicetree.h` before including `nvic.h`, so no problem there.
For `cmsis.h`, it includes `soc.h` before including `nvic.h`, which means `soc.h` (or one of the headers included by it) must include `devicetree.h`.

My question is: why have such a convoluted and indirect scheme when we can just `#include <devicetree.h>` in `nvic.h` as one would normally do? It seems `devicetree.h` only includes the headers that have one-way dependency and should never experience cross dependency issues.

Reply:
> Then I would insist on having SAM E5x SoCs include devicetree.h in soc.h. We started introducing this convention a while back; most of ARM SoCs already have transitioned to include devicetree.h in their soc.h. We should do it for the few ARM SoCs that are remaining.

I see that only about half of the SoCs seem to be following this convention. I do not necessarily object to this convention; but is this still relevant? (i.e. do we actually have a problem if we don't do this?) As far as I can see, `devicetree.h` is very well isolated and does not have any cross dependencies.

Reply:
@stephanosio in master branch, if you do not include devicetree.h in soc.h you cannot compile for user-mode. You can try this on E54.

This matter was identified 18 months back, after we eliminated a huge sort of inclusion cycles. Then we decided that by convention, all soc.h (at least for ARM) has to include devicetree.h (At that time, was called, generated_dts_board.h).

The SoCs that don't follow the convention yet are likely the ones that do not have MPU at all, or they don' have it yet enabled, such as the ATSAME54.

Now:

A week ago we added the nvic.h header. The reason was mostly to have NUM_PRIO_BITS included in assembly code. This header is included in cmsis.h which already soc.h, which already includes devicetree.h (or it has to, according to our convention). nvic.h header was never meant to change any inclusion paths! (its content could well be copy-pasted in the two places the header is included, or it could just be an internal-hidden header,. just to not duplicate the added content).

Including devicetree.h in nvic.h is not desired (but not wrong of course), because it hides the problem with SoCs not following the convention to include devicetree in the soc header.

Reply:
> in master branch, if you do not include devicetree.h in soc.h you cannot compile for user-mode. You can try this on E54.

I can successfully build and run `tests/kernel/mem_protect/userspace` and other userspace-enabled tests and samples on E54, which does not have `#include <devicetree.h>` in `soc.h` though.

Reply:
Maybe the recent `devicetree.h` rework has something to do with it? At least, for now, it no longer seems including `devicetree.h` in `soc.h` is a requirement.

Reply:
For now, I have added `devicetree.h` inclusion in all Atmel SAM(0) family SoCs, until we figure out if there is a better way to handle this after 2.3 release.